### PR TITLE
Add const checking for elements of tuple formals

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR adjusts the compiler to emit errors when a const element of a tuple formal is modified. Prior to this PR the compiler would compile programs such as the following without incident:

```chapel
record r { var x: int = 0; }

proc modifyDefaultIntentArg(tup) {
  tup[1] = new r(128);
}

proc test() {
  const a = (0, new r(0));
  modifyDefaultIntentArg(a);
  writeln(a);
}
test();
```

The call to `modifyDefaultIntentArg` incorrectly modifies the second element of the const tuple `r` via the formal argument `tup`. Because `tup` has the default intent, its second element should be passed by `const ref`, and any modification of the element within `modifyDefaultIntentArg` should be an error.

---

Add a function `checkTupleFormalUses` to `lateConstCheck`. This method checks non-ref/non-ref-if-modified elements of tuple formals to make sure their constness is not violated. If a field should be const but is set to `QUAL_REF` by `cullOverReferences`, then emit an error. The location that is reported is an imprecise guess - future work should attempt to make this reporting as precise as possible and provide more context for the use.

Add a set of `FnSymbol*` to the `for_formals_actuals` loop in `lateConstCheck` so that tuple formals are only checked once per function in `checkTupleFormalUses`, not once per call.

Add a function `checkTupleFormalToActual` to `lateConstCheck`. This method checks ref/ref-if-modified elements of tuple formals. If the element is ref or inferred to be ref and the actual element is const, emit an error. If the element is ref from ref-if-modified, provide an imprecise hint about where it was used - future work should make this reporting more precise.

Infer the concrete intent of blank tuple formals that contain single/sync/atomic elements as `INTENT_REF_MAYBE_CONST` instead of `INTENT_CONST_REF`. The blank intent for single/sync/atomics is `ref`. It does not make sense to infer the intent of the entire tuple as `INTENT_REF` if it contains such an element, so mimic what we do for blank intent tuples containing arrays.

Fix a bug in `cullOverReferences` function `CullRefCtx::collectTuplesAndRefMaybeConstArgs` that caused the compiler to incorrectly collect a reference to a tuple for analysis instead of the tuple itself.

Adjust `CullRefCtx::checkTupleCastCall` to only collect the LHS of a tuple cast call as a dependency if it contains ref elements. If the LHS is a value tuple, the compiler need not be concerned with how it is used.

In `CullRefCtx::checkAccessorLikeCall`, do not add the LHS as a dependency if it is known to be constant. Any misuse of a const LHS should have already been reported in a previous pass.

In `CullRefCtx::checkRefMaybeConstFormal`, escape early if the function being called has flag `FLAG_REF_TO_CONST_WHEN_CONST_THIS` and the actual is a const method receiver. Otherwise, we add a false dependency that the compiler may consider a "setting" use.

Add debugging tools to `lateConstCheck` in the spirit of `cullOverReferences`.

Clean up `lateConstCheck` by moving calls to skip into a function `isCallToSkip`. Additional "light" refactoring might help with integrating new code into the rest of the pass.

Adjust the IO `proc <~>` for writing channels so that the element to write is `const`.

Adjust the tests `types/tuple/dinan/pining_for_lisp.chpl` and `types/tuple/dinan/pining_for_lisp_ver2.chpl` so that their `car` and `cdr` functions return `const ref` instead of `ref`.

Add two tests to `types/tuple/const` that verify that the constness of tuple elements is respected for blank/const tuples for almost every element type. Future work should adjust the tests for tuples of sync/single when those are permitted in tuples. Future work should adjust the tests for tuples containing `dmap`.

---

Future work:

- Adjust `cullOverReferences` to store field index in addition to AST node in the `reasonNotConst` map, and then use this information to print more accurate error messages in `checkTupleFormalUses` and `checkTupleFormalToActual.
- Optionally, seek to integrate error reporting for tuples together with existing error reporting in `lateConstCheck`.
- Do const checking for tuples containing tuples.
- Add tests checking constness for tuples of `sync` and `single` when appropriate.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet`

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Drafting this until `COMM=none` passes.

**Comment by cassella** [comment]
In the PR description, do you mean `const tup =`, or to pass `r` to the function (and to writeln())?

**Comment by dlongnecke-cray** [comment]
Would adjusting the description to:

"modifies the second element of the const tuple r by assigning to the second element of tuple formal tup" help?

**Comment by dlongnecke-cray** [comment]
I see the error, thanks. I meant to pass `r` to the function.

**Comment by dlongnecke-cray** [comment]
Some errors to attend to over `release/examples`:

```
[Error compilation failed for release/examples/benchmarks/isx/isx]
[Error matching compiler output for release/examples/benchmarks/lulesh/lulesh (compopts: 1)][Error matching compiler output for release/examples/benchmarks/lulesh/lulesh (compopts: 2)]
[Error matching compiler output for release/examples/benchmarks/lulesh/test3DLulesh (compopts: 1)]
[Error matching compiler output for release/examples/benchmarks/lulesh/test3DLulesh (compopts: 2)]
[Error matching compiler output for release/examples/benchmarks/ssca2/SSCA2_main (compopts: 3)]
[Error matching compiler output for release/examples/benchmarks/ssca2/SSCA2_main (compopts: 4)]
[Error matching compiler output for release/examples/benchmarks/ssca2/SSCA2_main (compopts: 5)]
[Error matching compiler output for release/examples/primers/LinearAlgebralib]
[Error matching compiler output for release/examples/primers/learnChapelInYMinutes]
[Error matching compiler output for release/examples/primers/sparse]
[Error matching compiler output for release/examples/spec/Arrays/sparse-error]
[Error matching compiler output for release/examples/spec/Tuples/tuple-return-behavior]
```

**Comment by dlongnecke-cray** [comment]
Still ironing out a bunch of different bugs, this time due to iterators.

**Comment by dlongnecke-cray** [comment]
Down to just 2 failures, with mason:

```
[Error matching program output for mason/mason-external/libtomlc99/mason-external]
[Error matching program output for mason/mason-external/masonExternalRanges/mason-external-range]
```

**Comment by dlongnecke-cray** [comment]
Mason failures were not mine, but here are some actual failures to contend with: 

```
[Error matching compiler output for types/tuple/dinan/pining_for_lisp]
[Error matching compiler output for types/tuple/dinan/pining_for_lisp_ver2]
```

Need to adjust `cullOverReferences` some more to fix these, specifically `checkBuildTupleCalls` and `checkTupleCastCalls`.

**Comment by dlongnecke-cray** [comment]
All tests pass on `COMM=none`, now time to add checking for `REF_IF_MODIFIED` and `REF` tuple elements.

**Comment by dlongnecke-cray** [comment]
Checks for `REF` and `REF_MAYBE_CONST` elements have been added. Some new test failures to fix...

```
[Error matching compiler output for functions/ferguson/ref-pair/tuples/tuple-args1]
[Error matching compiler output for functions/ferguson/ref-pair/iterating-over-arrays]
[Error matching compiler output for functions/iterators/tzakian/atomic_pass_by_ref_leader_follower_iterator]
[Error matching compiler output for reductions/bradc/minmaxlocscan-shape]
[Error matching compiler output for reductions/bradc/minmaxlocscan]
[Error matching compiler output for types/records/split-init/tuple-split-init-out]
[Error compilation failed for types/tuple/const/testModifyArrArgConstActual]
[Error matching compiler output for types/tuple/const/tupleModifyBlankArgElem]
[Error compilation failed for types/tuple/const/tupleModifyBlankArgElemRefArr]
[Error matching compiler output for types/tuple/const/tupleModifyConstArgElem]
[Error matching compiler output for types/tuple/ferguson/tuple-array]
[Error matching compiler output for types/tuple/these/forall]
[Error matching compiler output for types/tuple/these/zip_forall]
```

**Comment by dlongnecke-cray** [comment]
Hi @vasslitvinov or @mppf, I think this effort is finally ready for review.

There are some small changes scattered around `cullOverReferences`. Please look at them and see if they seem reasonable.
Most of the work takes place in `lateConstCheck`. The new functions are not as integrated with old code as I would like, but that can remain as future work.

File: compiler/resolution/cullOverReferences.cpp [hunk 1]
```diff
@@ -1013,12 +1037,23 @@ bool CullRefCtx::checkAccessorLikeCall(SymExpr* se, CallExpr* call,
...
-        GraphNode srcNode = makeNode(lhsSymbol, node.fieldIndex);
-        collectedSymbols.push_back(srcNode);
-        addDependency(revisitGraph, srcNode, node);
-        revisit = true;
 
-        return true;
+        // Don't worry about tracking the destination if it's constant or
+        // marked with the flag FLAG_REF_TO_CONST.
+        if (lhsSymbol->hasFlag(FLAG_REF_TO_CONST) ||
+            lhsSymbol->isConstant() ||
```
Comment (mppf) [inline compiler/resolution/cullOverReferences.cpp:1044]: Huh, I wonder if this should call `lhsSymbol->isConstValWillNotChange()`.

File: compiler/resolution/lateConstCheck.cpp [hunk 2]
```diff
@@ -116,26 +123,520 @@ static FnSymbol* getSerialIterator(FnSymbol* fn) {
...
+    
+    // Check ref/ref-if-modified fields in "checkTupleFormalToActual".
+    if (fieldIntent == INTENT_REF_MAYBE_CONST ||
+        fieldIntent == INTENT_REF) {
+      INT_ASSERT(!isFormalIntentConst);
+      continue;
+    }
+
+    // The tuple intent is const, but it is used somewhere...
+    if (fieldIntent & INTENT_CONST && !isFieldMarkedConst) {
```
Comment (mppf) [inline compiler/resolution/lateConstCheck.cpp:256]: Put parens around the bitwise & operation here please

File: compiler/resolution/lateConstCheck.cpp [hunk 3]
```diff
@@ -116,26 +123,520 @@ static FnSymbol* getSerialIterator(FnSymbol* fn) {
...
+
+  AggregateType* at = toAggregateType(sym->getValType());
+  if (at == NULL || !at->symbol->hasFlag(FLAG_TUPLE)) {
+    return NULL;
+  }
+
+  // Find a PRIM_SET_MEMBER for the coerce temp. The third argument is
+  // the read field temporary.
+  //    ('.=' coerce_tmp x0 read_x0)
+  for_SymbolSymExprs(se, sym) {
```
Comment (mppf) [inline compiler/resolution/lateConstCheck.cpp:308]: It should probably check some element of the PRIM_SET_MEMBER is se

**Pull request merged**